### ISSUE TITLE
`mix credo` crashes when `@def` module attribute present

### ISSUE THREAD

**Issue by myronmarston** [rrrene/credo#83]
I just ran `mix credo` on my project (trying it out for the first time) and it crashed:

```
** (EXIT from #PID<0.47.0>) an exception was raised:
    ** (MatchError) no match of right hand side value: "Returns a list of `TimeSlice` structs based on the provided `time_slice_selector`.\n"
        lib/credo/code/parameters.ex:13: Credo.Code.Parameters.count/1
        lib/credo/priority.ex:77: Credo.Priority.priority_for/1
        lib/credo/priority.ex:59: Credo.Priority.traverse/2
        (elixir) lib/macro.ex:188: anonymous fn/4 in Macro.do_traverse/4
        (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
        (elixir) lib/macro.ex:187: Macro.do_traverse/4
        (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
        (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3

09:43:16.275 [error] Task #PID<0.1159.0> started from #PID<0.47.0> terminating
** (MatchError) no match of right hand side value: "Returns a list of `TimeSlice` structs based on the provided `time_slice_selector`.\n"
    lib/credo/code/parameters.ex:13: Credo.Code.Parameters.count/1
    lib/credo/priority.ex:77: Credo.Priority.priority_for/1
    lib/credo/priority.ex:59: Credo.Priority.traverse/2
    (elixir) lib/macro.ex:188: anonymous fn/4 in Macro.do_traverse/4
    (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    (elixir) lib/macro.ex:187: Macro.do_traverse/4
    (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
Function: #Function<8.7554626/0 in Credo.Check.Runner.run/2>
    Args: []

09:43:16.276 [error] GenServer Credo.Supervisor terminating
** (MatchError) no match of right hand side value: "Returns a list of `TimeSlice` structs based on the provided `time_slice_selector`.\n"
    lib/credo/code/parameters.ex:13: Credo.Code.Parameters.count/1
    lib/credo/priority.ex:77: Credo.Priority.priority_for/1
    lib/credo/priority.ex:59: Credo.Priority.traverse/2
    (elixir) lib/macro.ex:188: anonymous fn/4 in Macro.do_traverse/4
    (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    (elixir) lib/macro.ex:187: Macro.do_traverse/4
    (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
    (elixir) lib/enum.ex:1151: Enum."-map_reduce/3-lists^mapfoldl/2-0-"/3
```

The source of the problem was an `@def` module attribute:

``` elixir
  @def """
  Returns a list of `TimeSlice` structs based on the provided `time_slice_selector`.
  """
```

As it turns out, this was intended to be `@doc` (whoops!) so I guess Credo's crash helped identify the problem :).  But Credo should probably not crash due to the presence of particular module attributes.

**Comment by rrrene** [comment]
Mmmh, that's definitely a bug! Thanks for reporting, will fix! :+1:

**Comment by rrrene** [comment]
So it turns out Credo crashes when you define module attributes named `def`, `defp` or `defmacro`.

The reason for this is that we match on `{:def, meta, arguments}` inside a module's AST to find calls to the `def` macro (which defines functions, obviously).

If we define a `@def "This is a string"` module attribute we get the following AST

    {:@, [line: 1], [{:def, [line: 1], ["This is a string"]}]}

which contains the part we are matching for

    {:def, [line: 1], ["This is a string"]}

To summarize: I will probably have to patch all checks to safeguard against this. I'm glad you found this.
